[logging] Log cudagraphify timings to dynamo_timed

masnesral opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #143220

Summary: this adds some new dynamo_timed calls in cudagraph_trees, primarily with the aim to add cudagraph-related timing to scuba. Things to note:
* Uses the changes in https://github.com/pytorch/pytorch/pull/141919 to log "runtime" entries
* The logging for chromium/tlparse/scuba relies on us providing a compile_id since it's not available in the environment. A lot of the changes here are just passing around the compile_id
* I believe the spirit of the scuba logging is to capture the overheads of `torch.compile`. Therefore, I'm not adding _every_ dynamo_timed to scuba. For example, "run_eager" is the first real execution of the inductor graph -- it's not cudagraph overhead, per se. Watch out for the two instances of `dynamo_compile_runtime_column_us="runtime_cudagraphify_time_us"`. Those are the spots I believe are _extra_ overhead we'd contribute to torch.compile.

Test Plan:
`python benchmarks/dynamo/torchbench.py --performance --training --amp --backend inductor --device cuda --print-compilation-time --repeat 5 --cold-start-latency --only dcgan`:
* tlparse: https://fburl.com/21yrdn8h
* scuba: https://fburl.com/scuba/dynamo_compile/sandbox/wt90wnjz

`python benchmarks/dynamo/torchbench.py --performance --training --amp --backend inductor --device cuda --print-compilation-time --repeat 5 --cold-start-latency --only nanogpt`
* tlparse: https://fburl.com/r9mp7uiv
* scuba: https://fburl.com/scuba/dynamo_compile/sandbox/1nvx94re

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @ColinPeppler @amjames @desertfire @chauhang @aakhundov

masnesral committed:
[logging] Log cudagraphify timings to dynamo_timed

Test Plan: Will provide scuba logs once agree on what we want to count towards overall cudagraph overhead

[ghstack-poisoned]

masnesral commented on torch/_inductor/compile_fx.py:
```diff
@@ -1199,10 +1200,7 @@ def cudagraphify(
-                "cudagraphify",
-                log_pt2_compile_event=True,
-            ), dynamo_utils.preserve_rng_state():
+            with dynamo_utils.preserve_rng_state():
```
Talk to @BoyuanFeng ; this dynamo_timed wasn't giving us anything.

masnesral commented on torch/_inductor/cudagraph_trees.py:
```diff
@@ -416,25 +417,43 @@ def cudagraphify(
+    with dynamo_timed(
+        "cudagraphify.get_container",
+        log_pt2_compile_event=True,
+        # This operation is strictly cudagraph overhead, so we count it
```
This part is important. @ezyang, can you vet the rationale in the summary? I talked to @eellison and I'm trying to capture only the _additional_ overhead of cudagraphs, e.g., the call to `run_eager` elsewhere does not get added to the column `runtime_cudagraphify_time_us`. Check out the perfetto traces linked in the summary to see what I mean.

Also something to note: experimentally, the overhead you'll see in the sample perfetto traces is coming from here: https://github.com/pytorch/pytorch/blob/bb574abe73aa84fd9b895b4c7c39ad79bd53f037/torch/_inductor/cudagraph_trees.py#L1874
We could wrap the dynamo_timed around that specifically, but it'd involve passing the compile_id through a bunch of stuff and capturing the timing here seems almost as good.

masnesral commented on torch/_inductor/cudagraph_trees.py:
```diff
@@ -416,25 +417,43 @@ def cudagraphify(
+    ):
+        manager = get_container(device_index).get_tree_manager()
+
+    with dynamo_timed(
```
See sample perfetto traces. It's not entirely clear we need this one.

masnesral commented on torch/_inductor/cudagraph_trees.py:
```diff
@@ -981,9 +1002,18 @@ def maybe_get_static_data_ptr(
+            log_pt2_compile_event=True,
+            # This operation is strictly cudagraph overhead, so we count it
+            # towards the overall cudagraphify time when logging to scuba.
+            dynamo_compile_runtime_column_us="runtime_cudagraphify_time_us",
```
We're also adding this to the scuba column, but not the replay below

eellison commented on torch/_inductor/cudagraph_trees.py:
```diff
@@ -2008,7 +2050,13 @@ def exceed_rerecord_limit(
+        self,
+        new_inputs: List[InputType],
+        function_id: FunctionID,
+        mode: CompilationMode,
```
Mode is mapped by `self.id_to_mode`, shouldn't have to pass in explicitly. can we do something similar with function_id as well ?

eellison reviewed (commented):
Thanks for doing this, looks good! Could we try to refactor a bit ?

masnesral commented on torch/_inductor/cudagraph_trees.py:
```diff
@@ -416,25 +417,43 @@ def cudagraphify(
-        placeholders,
-        mutated_input_idxs,
-    )
+    with dynamo_timed(
```
I think you mean so that we can use as a decorator? Unfortunately it looks like there's some unfortunate interaction with dynamo_timed and cProfile when used as a cm. We have this warning: https://github.com/pytorch/pytorch/blob/91bf2e16debdc41f5dde2bb5cc8e4f39f8955d4e/torch/_dynamo/utils.py#L334

masnesral replied:
I could define a new context manager?

```
@contextmanager
def dynamo_timed_cudagraphs(name, compile_id, mode, overhead=False):
    with dynamo_timed(
        name,
        log_pt2_compile_event=True,
        compile_id=compile_id,
        is_forward=mode != CompilationMode.BACKWARD,
       dynamo_compile_runtime_column_us= runtime_cudagraphify_time_us" if overhead else None, 
    ):
        yield
```

That'd make the usages a bit more concise ... wdyt?

eellison replied:
Sure, sgtm!

masnesral committed:
Update on "[logging] Log cudagraphify timings to dynamo_timed"


Summary: this adds some new dynamo_timed calls in cudagraph_trees, primarily with the aim to add cudagraph-related timing to scuba. Things to note:
* Uses the changes in https://github.com/pytorch/pytorch/pull/141919 to log "runtime" entries
* The logging for chromium/tlparse/scuba relies on us providing a compile_id since it's not available in the environment. A lot of the changes here are just passing around the compile_id
* I believe the spirit of the scuba logging is to capture the overheads of `torch.compile`. Therefore, I'm not adding _every_ dynamo_timed to scuba. For example, "run_eager" is the first real execution of the inductor graph -- it's not cudagraph overhead, per se. Watch out for the two instances of `dynamo_compile_runtime_column_us="runtime_cudagraphify_time_us"`. Those are the spots I believe are _extra_ overhead we'd contribute to torch.compile.

Test Plan:
`python benchmarks/dynamo/torchbench.py --performance --training --amp --backend inductor --device cuda --print-compilation-time --repeat 5 --cold-start-latency --only dcgan`:
* Perfetto: https://fburl.com/jr8k587c
* Scuba: https://fburl.com/scuba/dynamo_compile/sandbox/fstyrsn4

`python benchmarks/dynamo/torchbench.py --performance --training --amp --backend inductor --device cuda --print-compilation-time --repeat 5 --cold-start-latency --only nanogpt`
* Perfetto: https://fburl.com/v5uo17eb
* Scuba: https://fburl.com/scuba/dynamo_compile/sandbox/wzdsxd45

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang aakhundov

[ghstack-poisoned]

masnesral commented:
@pytorchbot merge

masnesral committed:
Update on "[logging] Log cudagraphify timings to dynamo_timed"


Summary: this adds some new dynamo_timed calls in cudagraph_trees, primarily with the aim to add cudagraph-related timing to scuba. Things to note:
* Uses the changes in https://github.com/pytorch/pytorch/pull/141919 to log "runtime" entries
* The logging for chromium/tlparse/scuba relies on us providing a compile_id since it's not available in the environment. A lot of the changes here are just passing around the compile_id
* I believe the spirit of the scuba logging is to capture the overheads of `torch.compile`. Therefore, I'm not adding _every_ dynamo_timed to scuba. For example, "run_eager" is the first real execution of the inductor graph -- it's not cudagraph overhead, per se. Watch out for the two instances of `dynamo_compile_runtime_column_us="runtime_cudagraphify_time_us"`. Those are the spots I believe are _extra_ overhead we'd contribute to torch.compile.

Test Plan:
`python benchmarks/dynamo/torchbench.py --performance --training --amp --backend inductor --device cuda --print-compilation-time --repeat 5 --cold-start-latency --only dcgan`:
* Perfetto: https://fburl.com/jr8k587c
* Scuba: https://fburl.com/scuba/dynamo_compile/sandbox/fstyrsn4

`python benchmarks/dynamo/torchbench.py --performance --training --amp --backend inductor --device cuda --print-compilation-time --repeat 5 --cold-start-latency --only nanogpt`
* Perfetto: https://fburl.com/v5uo17eb
* Scuba: https://fburl.com/scuba/dynamo_compile/sandbox/wzdsxd45

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang aakhundov

[ghstack-poisoned]

masnesral committed:
Update on "[logging] Log cudagraphify timings to dynamo_timed"


Summary: this adds some new dynamo_timed calls in cudagraph_trees, primarily with the aim to add cudagraph-related timing to scuba. Things to note:
* Uses the changes in https://github.com/pytorch/pytorch/pull/141919 to log "runtime" entries
* The logging for chromium/tlparse/scuba relies on us providing a compile_id since it's not available in the environment. A lot of the changes here are just passing around the compile_id
* I believe the spirit of the scuba logging is to capture the overheads of `torch.compile`. Therefore, I'm not adding _every_ dynamo_timed to scuba. For example, "run_eager" is the first real execution of the inductor graph -- it's not cudagraph overhead, per se. Watch out for the two instances of `dynamo_compile_runtime_column_us="runtime_cudagraphify_time_us"`. Those are the spots I believe are _extra_ overhead we'd contribute to torch.compile.

Test Plan:
`python benchmarks/dynamo/torchbench.py --performance --training --amp --backend inductor --device cuda --print-compilation-time --repeat 5 --cold-start-latency --only dcgan`:
* tlparse: https://fburl.com/21yrdn8h
* scuba: https://fburl.com/scuba/dynamo_compile/sandbox/wt90wnjz

`python benchmarks/dynamo/torchbench.py --performance --training --amp --backend inductor --device cuda --print-compilation-time --repeat 5 --cold-start-latency --only nanogpt`
* tlparse: https://fburl.com/r9mp7uiv
* scuba: https://fburl.com/scuba/dynamo_compile/sandbox/1nvx94re

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang aakhundov

[ghstack-poisoned]

masnesral committed:
Update on "[logging] Log cudagraphify timings to dynamo_timed"


Summary: this adds some new dynamo_timed calls in cudagraph_trees, primarily with the aim to add cudagraph-related timing to scuba. Things to note:
* Uses the changes in https://github.com/pytorch/pytorch/pull/141919 to log "runtime" entries
* The logging for chromium/tlparse/scuba relies on us providing a compile_id since it's not available in the environment. A lot of the changes here are just passing around the compile_id
* I believe the spirit of the scuba logging is to capture the overheads of `torch.compile`. Therefore, I'm not adding _every_ dynamo_timed to scuba. For example, "run_eager" is the first real execution of the inductor graph -- it's not cudagraph overhead, per se. Watch out for the two instances of `dynamo_compile_runtime_column_us="runtime_cudagraphify_time_us"`. Those are the spots I believe are _extra_ overhead we'd contribute to torch.compile.

Test Plan:
`python benchmarks/dynamo/torchbench.py --performance --training --amp --backend inductor --device cuda --print-compilation-time --repeat 5 --cold-start-latency --only dcgan`:
* tlparse: https://fburl.com/21yrdn8h
* scuba: https://fburl.com/scuba/dynamo_compile/sandbox/wt90wnjz

`python benchmarks/dynamo/torchbench.py --performance --training --amp --backend inductor --device cuda --print-compilation-time --repeat 5 --cold-start-latency --only nanogpt`
* tlparse: https://fburl.com/r9mp7uiv
* scuba: https://fburl.com/scuba/dynamo_compile/sandbox/1nvx94re

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang aakhundov

[ghstack-poisoned]

masnesral commented:
@pytorchbot merge
